zend_abstract_interface/functions/php8/functions.c

```diff
@@ -0,0 +1,139 @@
+#include "../functions.h"
+
+#include <assert.h>
+#include <sandbox/sandbox.h>
+
+#ifndef NDEBUG
+static bool z_is_lower(const char *str) {
+    char *p = (char *)str;
+    while (*p) {
+        if (isalpha(*p) && !islower(*p)) return false;
+        p++;
+    }
+    return true;
+}
+#endif
+
+bool zai_call_function(const char *name, size_t name_len, zval *retval, int argc, ...) {
+    if (!retval) return false;
+
+    /* For consistency with zend_call_function(), this wrapper also initializes
+     * the retval to IS_UNDEF so that the caller can expect an undefined retval
+     * in the error cases. Past wrappers of zend_call_function() did not mimic
+     * this behavior and that lead to a SIGSEGV.
+     *
+     * https://github.com/php/php-src/blob/PHP-8.0.3/Zend/zend_execute_API.c#L673
+     */
+    ZVAL_UNDEF(retval);
+
+    if (!name || !name_len) return false;
+
+    /* Functions cannot be called outside of a request context.
+     * PG(modules_activated) indicates that all of the module RINITs have been
+     * called and we are in a request context.
+     */
+    if (!PG(modules_activated)) return false;
+
+    assert(z_is_lower(name) && "Function names must be lowercase.");
```

Comment:
> On the other hand - unless PHP does some magic that I am not aware of - assert would cause the program execution to abort.

This is only true for debug builds of PHP because [`NDEBUG` is defined for non-debug builds of PHP](https://github.com/php/php-src/blob/c40231afbf85e0828a6c56d289cb823d65d03a54/main/php.h#L120-L126). To validate this, we [run the ZAI tests on non-debug builds](https://github.com/DataDog/dd-trace-php/blob/6f1f4b2/.circleci/config.yml#L1083) to ensure there is no abort happening for [these assertions](https://github.com/DataDog/dd-trace-php/blob/6f1f4b2/zend_abstract_interface/methods/tests/php5/methods.cc#L63-L87).

> Since the user is already "forced" to handle failure due to the return false above, do you think it would make sense here to just return false here are well?

This is the behavior on non-debug builds of PHP since the hash table lookup will fail if the function name is not lowercase (and then return `false`).